File: src/main/java/com/jackmeng/stl/stl_Monad.java
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package com.jackmeng.stl;

public interface stl_Monad< T >
{
    < R > stl_Monad< R > bind(stl_Function< ? super T, stl_Monad< R > > f);

    static < T > stl_Monad< T > unit(T value)
    {
        return new stl_Monad<>() {
            @Override public < R > stl_Monad< R > bind(stl_Function< ? super T, stl_Monad< R > > f)
            {
                return f.apply(value);
            }
        };
    }
}

File: src/main/java/com/jackmeng/stl/stl_Monad.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

interface stl_Monad<T>
{
	fun <R> bind(f:stl_Function<in T? , stl_Monad<R>?>):stl_Monad<R>?
	
	companion object
	{
		fun <T> unit(value:T):stl_Monad<T>
		{
			return object:stl_Monad<T>
			{
				override fun <R> bind(f:stl_Function<in T? , stl_Monad<R>?>):stl_Monad<R>?
				{
					return f.apply(value)
				}
			}
		}
	}
}